### PR TITLE
🏗 Add a `--local_changes` mode to `gulp prettify`

### PR DESCRIPTION
**Pull request by rsimha**
**PR highlights:**

- Adds a `--local_changes` mode to `gulp prettify`
- Uses the [`globby`](https://www.npmjs.com/package/globby) package to convert an array of globs to a list of matching files (fast, widely used, actively maintained, can handle negative globs)
- Refactors `lintableFilesChanged()` from `lint.js` into `getFilesChanged(globs)` in `build-system/common/utils.js`
- Makes `gulp pr-check` run `gulp prettify --local_changes`
- Makes the default pre-push hook run `gulp prettify --local_changes`

**Coming up:**
- Stop using `globs-to-files` across `build-system/` (hasn't been updated in 4 years / can't handle negative globs) and replace it with `globby`

Follow up to #25057 and #25085

### CODE REVIEW

File: build-system/common/utils.js [hunk 1]
```diff
@@ -30,6 +33,31 @@ function logOnSameLine(message) {
...
 }
 
+/**
+ * Converts an array of globs to a list of matching files using gulp.src, which
+ * can handle negative globs.
+ *
+ * @param {!Array<string>} globs
+ * @return {!Array<string>}
+ */
+async function globsToFiles(globs) {
```
Comment (danielrozenberg) [inline build-system/common/utils.js]: What do you mean by "negative globs"? We use the `globs-to-files` npm package in other places in the code, does that not work here?

File: build-system/tasks/prettify.js [hunk 2]
```diff
@@ -48,43 +49,88 @@ const PrettierResult = {
...
 
 /**
- * Checks files for formatting (and optionally fixes them) with Prettier.
- * Returns the cumulative result to the `gulp` process via  process.exitCode so
- * that all files can be checked / fixed.
+ * Extracts the list of prettifiable files in this PR from the commit log.
+ *
+ * @return {!Array<string>}
+ */
+async function getFilesChanged() {
```
Comment (estherkim) [inline build-system/tasks/prettify.js]: This is pretty similar to gulp lint's `lintableFilesChanged()`. Is there any overlap between the two that would be worth moving to `build-system/common/utils.js`?

File: build-system/tasks/prettify.js [hunk 3]
```diff
@@ -48,43 +49,88 @@ const PrettierResult = {
...
     });
+  }
+}
+
+/**
+ * Checks files for formatting (and optionally fixes them) with Prettier.
+ */
+async function prettify() {
+  maybeUpdatePackages();
+  let filesToCheck = await globsToFiles(prettifyGlobs);
```
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/prettify.js]: Good idea. Done.

File: build-system/common/utils.js (hunk 1 above)
Comment (rsimha) [reply build-system/common/utils.js]: > What do you mean by "negative globs"?

Something like `'!**/{node_modules,build,dist,dist.3p,dist.tools}/**'`.

> We use the `globs-to-files` npm package in other places in the code, does that not work here?

It doesn't. See "coming up" section in PR description.

File: build-system/tasks/prettify.js (hunk 2 above)
Comment (rsimha) [reply build-system/tasks/prettify.js]: You're right. I did spend some time trying to unify the two, but they currently use different globbing mechanisms and a different filter. Okay to re-attempt this in a future PR that refactors this and other tooling functionality into `build-system/common/utils.js`?

Comment (rsimha) [reply build-system/tasks/prettify.js]: Wait, now that I read it again, I think it's reasonable to do it in this PR. Coming up!

File: build-system/common/utils.js (hunk 1 above)
Comment (danielrozenberg) [reply build-system/common/utils.js]: Note to self: also read PR descriptions 🙃

Comment (rsimha) [reply build-system/common/utils.js]: Your comment was useful, in that it made me re-evaluate the wisdom of hacking together this implementation :)

A better and faster alternative that is being actively maintained is [`globby`](https://www.npmjs.com/package/globby). It supports everything we need, and can replace the function I wrote (and eventually, all the `globs-to-files` calls).

File: build-system/tasks/prettify.js (hunk 2 above)
Comment (rsimha) [reply build-system/tasks/prettify.js]: Done in the latest commit.

File: build-system/common/utils.js (hunk 1 above)
Comment (rsimha) [reply build-system/common/utils.js]: Done in the latest commit.

**Comment by rsimha** [comment]
@estherkim Good call on refactoring `lintableFilesChanged()`. This is ready for final review.

**Pull request merged**